File: jbake-core/src/test/java/org/jbake/render/SitemapRendererTest.java
package org.jbake.render;

import org.jbake.app.ContentStore;
import org.jbake.app.Renderer;
import org.jbake.app.configuration.DefaultJBakeConfiguration;
import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.template.RenderingException;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class SitemapRendererTest {

    @Test
    public void returnsZeroWhenConfigDoesNotRenderSitemaps() throws RenderingException {
        SitemapRenderer renderer = new SitemapRenderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderSiteMap()).thenReturn(false);

        ContentStore contentStore = mock(ContentStore.class);

        Renderer mockRenderer = mock(Renderer.class);
        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        assertThat(renderResponse).isEqualTo(0);
    }

    @Test
    public void doesNotRenderWhenConfigDoesNotRenderSitemaps() throws Exception {
        SitemapRenderer renderer = new SitemapRenderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderSiteMap()).thenReturn(false);

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, never()).renderSitemap(anyString());
    }

    @Test
    public void returnsOneWhenConfigRendersSitemaps() throws RenderingException {
        SitemapRenderer renderer = new SitemapRenderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderSiteMap()).thenReturn(true);

        ContentStore contentStore = mock(ContentStore.class);

        Renderer mockRenderer = mock(Renderer.class);

        int renderResponse = renderer.render(mockRenderer, contentStore, configuration);

        assertThat(renderResponse).isEqualTo(1);
    }

    @Test
    public void doesRenderWhenConfigDoesRenderSitemaps() throws Exception {
        SitemapRenderer renderer = new SitemapRenderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderSiteMap()).thenReturn(true);
        when(configuration.getSiteMapFileName()).thenReturn("mocksitemap.html");

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, times(1)).renderSitemap(anyString());
    }

    @Test(expected = RenderingException.class)
    public void propogatesRenderingException() throws Exception {
        SitemapRenderer renderer = new SitemapRenderer();

        JBakeConfiguration configuration = mock(DefaultJBakeConfiguration.class);
        when(configuration.getRenderSiteMap()).thenReturn(true);
        when(configuration.getSiteMapFileName()).thenReturn("mocksitemap.html");

        ContentStore contentStore = mock(ContentStore.class);
        Renderer mockRenderer = mock(Renderer.class);

        doThrow(new Exception()).when(mockRenderer).renderSitemap(anyString());

        renderer.render(mockRenderer, contentStore, configuration);

        verify(mockRenderer, never()).renderSitemap(anyString());
    }

}

File: jbake-core/src/test/kotlin/org/jbake/render/SitemapRendererTest.kt
package org.jbake.render

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.jbake.TestUtils
import org.jbake.app.ContentStore
import org.jbake.app.ContentStoreIntegrationTest.Companion.tempDir
import org.jbake.app.Renderer
import org.jbake.app.RenderingException
import org.jbake.app.configuration.DefaultJBakeConfiguration

class SitemapRendererTest : StringSpec({
    "returnsZeroWhenRendersSitemaps" {
        val tool = SitemapRenderingTool()

        val configuration = mockk<DefaultJBakeConfiguration>()
        every { configuration.renderSiteMap } returns false
        val contentStore = mockk<ContentStore>()
        val mockRenderer = mockk<Renderer>()

        val renderResponse = tool.render(mockRenderer, contentStore, configuration)
        renderResponse shouldBe 0
    }

    "doesNotRenderWhenConfigDoesNotRenderSitemaps" {
        val tool = SitemapRenderingTool()

        val configuration = mockk<DefaultJBakeConfiguration>()
        every { configuration.renderSiteMap } returns false
        every { configuration.siteMapFileName } returns "sitemap.xml"
        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        val mockRenderer = Renderer(contentStore, configuration, renderingEngine)

        tool.render(mockRenderer, contentStore, configuration)

        // No verification needed - we just check it doesn't throw
    }

    "returnsOneWh0enRendersSitemaps" {
        val tool = SitemapRenderingTool()

        val configuration = mockk<DefaultJBakeConfiguration>()
        every { configuration.renderSiteMap } returns true
        val contentStore = mockk<ContentStore>()
        val mockRenderer = mockk<Renderer>(relaxed = true)
        every { mockRenderer.renderSitemap() } returns Unit

        val renderResponse = tool.render(mockRenderer, contentStore, configuration)
        renderResponse shouldBe 1
    }

    "rendersSitemapsWhenConfigured" {
        val tool = SitemapRenderingTool()

        val configuration = mockk<DefaultJBakeConfiguration>(relaxed = true)
        every { configuration.renderSiteMap } returns true
        every { configuration.siteMapFileName } returns "mocksitemap.html"
        every { configuration.destinationDir } returns TestUtils.createOrEmptyDir(tempDir, "output")
        every { configuration.renderEncoding } returns "UTF-8"
        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        val renderer = Renderer(contentStore, configuration, renderingEngine)

        val result = tool.render(renderer, contentStore, configuration)

        result shouldBe 1
        verify(exactly = 1) { renderingEngine.renderDocument(any(), any(), any()) }
    }

    "propagatesRenderingException" {
        val tool = SitemapRenderingTool()

        val configuration = mockk<DefaultJBakeConfiguration>(relaxed = true)
        every { configuration.renderSiteMap } returns true
        every { configuration.siteMapFileName } returns "mocksitemap.html"
        every { configuration.destinationDir } returns TestUtils.createOrEmptyDir(tempDir, "output")
        val contentStore = mockk<ContentStore>(relaxed = true)
        val renderingEngine = mockk<org.jbake.template.DelegatingTemplateEngine>(relaxed = true)
        every { renderingEngine.renderDocument(any(), any(), any()) } throws RuntimeException("Test exception")
        val renderer = Renderer(contentStore, configuration, renderingEngine)

        shouldThrow<RenderingException> {
            tool.render(renderer, contentStore, configuration)
        }
    }
})
